File: leetcode_1305/cpp/leetcode_1305.cpp
//
// Created by wangzy on 2022/4/9.
//

#include<iostream>
#include<queue>
#include<vector>
#include <algorithm>
#include <unordered_map>
#include <list>
#include <deque>
#include <stack>
#include <sstream>
#include <typeinfo>
#include <climits>
#include <random>
#include <ctime>

using namespace std;

struct TreeNode{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0),left(nullptr),right(nullptr) {}
    TreeNode(int x) : val(x),left(nullptr),right(nullptr) {}
    TreeNode(int x,TreeNode *left,TreeNode* right):val(x),left(left),right(right) {}
};

class Solution {
public:
    vector<int> getAllElements(TreeNode* root1, TreeNode* root2) {
        //思路：1、利用中序遍历将BST的节点以从小到大顺序排列
        //2、再利用数组排序的方式排序两个列表
        vector<int> vec1;
        vector<int> vec2;
        inOrderBSTHelper(root1,vec1);
        inOrderBSTHelper(root2,vec2);
        vector<int> res;
        sortArrHelper(vec1,vec2,res);
        return res;
    }
    vector<string> printTree(TreeNode* root){
        string s;
        vector<string> levelOrder = levelOrderStr(root);
        return levelOrder;
    }
    void deleteTree(TreeNode* root){
        if(root){
            deleteTree(root->left);
            deleteTree(root->right);
            root = nullptr;
        }
    }
    vector<string> levelOrderStr(TreeNode* root){
        vector<TreeNode*> cur;
        vector<TreeNode*> next;
        if(!root)
            return vector<string> {};
        if(root)
            cur.push_back(root);
        vector<string> res;
//        vector<string> curEles;
        while(cur.size()){
            //标识next是否都为null
            int flag = 0;
            for(int i = 0;i < cur.size();i++){
                TreeNode* node = cur[i];
                if(!node){
                    res.push_back("null");
                    continue;
                }
                else
                    res.push_back(to_string(node->val));
                //如果node的孩子不为null,则改变flag
                if(node->left || node->right)
                    flag++;
                next.push_back(node->left);
                next.push_back(node->right);
            }
            cur = next;
            if(flag == 0)
                cur.clear();
            next.clear();
        }
        return res;
    }
private:
    void inOrderBSTHelper(TreeNode* root,vector<int>& vec)    {
        if(root){
            inOrderBSTHelper(root->left,vec);
            vec.push_back(root->val);
            inOrderBSTHelper(root->right,vec);
        }
    }
    void sortArrHelper(vector<int>& vec1,vector<int>& vec2,vector<int>& res){
        int idx1 = 0;int idx2 = 0;
        while(idx1 < vec1.size() && idx2 < vec2.size()){
            if(vec1[idx1] <= vec2[idx2])
                res.push_back(vec1[idx1++]);
            else //vec1[idx1] > vec2[idx2]
                res.push_back(vec2[idx2++]);
        }
        if(idx1 == vec1.size()){
            while(idx2 < vec2.size())
                res.push_back(vec2[idx2++]);
        }else{
            //idx2 == vec2.size()
            while(idx1 < vec1.size())
                res.push_back(vec1[idx1++]);
        }
    }
};

template<typename T>
string convertVecToStr(std::vector<T> l){
    std::string res = "[";
    std::string tmp = "";
    std::stringstream ss;
    for(auto ele : l){
        ss << ele;
        ss >> tmp;
        res += tmp + ",";
        ss.clear();
    }
    res = res.substr(0,res.size()-1);
    res += "]";
    return res;
}

template<typename T>
void printVecArr(std::vector<std::vector<T>> l){
    std::string res = "[";
    std::string tmp = "";
    std::stringstream ss;
    for(auto ele : l){
        ss << convertVecToStr(ele);
        ss >> tmp;
        res += tmp + ",";
        ss.clear();
    }
    res = res.substr(0,res.size()-1);
    res += "]";
    std::cout<<res<<std::endl;
}

template<typename T>
void printVec(std::vector<T> l){
    std::string res = "[";
    std::string tmp = "";
    std::stringstream ss;
    for(auto ele : l){
        ss << ele;
        ss >> tmp;
        res += tmp + ",";
        ss.clear();
    }
    res = res.substr(0,res.size()-1);
    res += "]";
    std::cout<<res<<std::endl;
}

int main() {
    TreeNode *root1 = new TreeNode(2);
    TreeNode *left1 = new TreeNode(1);
    TreeNode *right1 = new TreeNode(4);
    TreeNode *root2 = new TreeNode(1);
    TreeNode *left2 = new TreeNode(0);
    TreeNode *right2 = new TreeNode(3);
//    TreeNode *ll = new TreeNode(1);
//    TreeNode *lr = new TreeNode(7);
//    TreeNode *rl = nullptr;
//    TreeNode *rr = new TreeNode(3);
    root1->left = left1;
    root1->right = right1;
    root2->left = left2;
    root2->right = right2;
//    left->left = ll;
//    left->right = lr;
//    right->left = rl;
//    right->right = rr;
//    ll->left = new TreeNode(0);
//    rr->right = new TreeNode(4);
    vector<string> rootArr1 = Solution().levelOrderStr(root1);
    printVec(rootArr1);
    //root1 = [2,1,4]
    vector<string> rootArr2 = Solution().levelOrderStr(root2);
    printVec(rootArr2);
    //root2 = [1,0,3]
    vector<int> res = Solution().getAllElements(root1,root2);
    printVec(res);
    //result
    //[0,1,1,2,3,4]
}

File: leetcode_1305/java/leetcode_1305.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;
public class leetcode_1305 {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    class Solution {
        public List<Integer> getAllElements(TreeNode root1, TreeNode root2) {
            //思路：1、利用中序遍历将BST的节点以从小到大顺序排列
            //2、再利用数组排序的方式排序两个列表
            List<Integer> list1 = new ArrayList();
            List<Integer> list2 = new ArrayList();
            List<Integer> res = new LinkedList();
            inOrder(root1,list1);
            inOrder(root2,list2);
            sortArrHelper(list1,list2,res);
            return res;
        }
        private void inOrder(TreeNode root,List<Integer> list){
            if(root != null){
                inOrder(root.left,list);
                list.add(root.val);
                inOrder(root.right,list);
            }
        }
        private void sortArrHelper(List<Integer> list1,List<Integer> list2,List<Integer> res){
            int idx1 = 0,idx2 = 0;
            while(idx1 < list1.size() && idx2 < list2.size()){
                if(list1.get(idx1).compareTo(list2.get(idx2)) <= 0)
                    res.add(new Integer(list1.get(idx1++)));
                else //list1.get(idx1).compareTo(list2.get(idx2)) > 0
                    res.add(new Integer(list2.get(idx2++)));
            }
            if(idx1 == list1.size()){
                while(idx2 < list2.size())
                    res.add(new Integer(list2.get(idx2++)));
            }else{
                //idx2 == list2.size()
                while(idx1 < list1.size())
                    res.add(new Integer(list1.get(idx1++)));
            }
        }
        public List<List<Integer>> levelOrder(TreeNode root) {
            List<TreeNode> cur = new ArrayList();
            List<TreeNode> next = new ArrayList();
            if (root == null)
                return new LinkedList();
            if (root != null)
                cur.add(root);
            List<List<Integer>> res = new LinkedList();
            List<Integer> curEles = new LinkedList();
            while (cur.size() > 0) {
                for (int i = 0; i < cur.size(); i++) {
                    curEles.add(cur.get(i).val);
                    if (cur.get(i).left != null)
                        next.add(cur.get(i).left);
                    if (cur.get(i).right != null)
                        next.add(cur.get(i).right);
                }
                res.add(new LinkedList(curEles));
                cur = new LinkedList(next);
                curEles.clear();
                next.clear();
            }
            return res;
        }

        public List<String> levelOrderStr(TreeNode root) {
            List<TreeNode> cur = new ArrayList();
            List<TreeNode> next = new ArrayList();
            if (root == null)
                return new LinkedList();
            if (root != null)
                cur.add(root);
            List<String> res = new LinkedList();
            while (cur.size() > 0) {
                int flag = 0;
                for (int i = 0; i < cur.size(); i++) {
                    if (cur.get(i) == null) {
                        res.add("null");
                        continue;
                    } else
                        res.add(String.valueOf(cur.get(i).val));
                    TreeNode left = cur.get(i).left;
                    TreeNode right = cur.get(i).right;
                    if (left != null || right != null)
                        flag++;
                    next.add(left);
                    next.add(right);
                }
                cur = new LinkedList(next);
                if (flag == 0)
                    cur.clear();
                next.clear();
            }
            return res;
        }

        public <T> void printListArr(List<List<T>> list) {
            for (int i = 0; i < list.size(); i++) {
                if (i == list.size() - 1) {
                    if (i == 0)
                        System.out.printf("[ ");
                    System.out.println(Arrays.toString(list.get(i).toArray((T[]) new Object[0])) + " ] ");
                } else {
                    if (i == 0)
                        System.out.printf("[ ");
                    System.out.printf(Arrays.toString(list.get(i).toArray((T[]) new Object[0])) + " , ");
                }
            }
        }
    }
    public static void main(String[] args) {
        leetcode_1305.Solution solution = new leetcode_1305().new Solution();
        leetcode_1305.TreeNode root1 = new leetcode_1305().new TreeNode(2);
        leetcode_1305.TreeNode left1 = new leetcode_1305().new TreeNode(1);
        leetcode_1305.TreeNode right1 = new leetcode_1305().new TreeNode(4);
        //root1 = [2,1,4]
        root1.left = left1;
        root1.right = right1;
        leetcode_1305.TreeNode root2 = new leetcode_1305().new TreeNode(1);
        leetcode_1305.TreeNode left2 = new leetcode_1305().new TreeNode(0);
        leetcode_1305.TreeNode right2 = new leetcode_1305().new TreeNode(3);
        //root2 = [1,0,3]
        root2.left = left2;
        root2.right = right2;
        List<String> root1Str = solution.levelOrderStr(root1);
        System.out.println(Arrays.toString(root1Str.toArray(String[]::new)));
        List<String> root2Str = solution.levelOrderStr(root2);
        System.out.println(Arrays.toString(root2Str.toArray(String[]::new)));
        List<Integer> allElements = solution.getAllElements(root1, root2);
        System.out.println(Arrays.toString(allElements.toArray((T[]) new Object[0])));
        //[0,1,1,2,3,4]
    }
}
